In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning)


In [ ]:
import pandas as pd
from google.colab import drive

# Kết nối Google Drive
drive.mount('/content/drive')

# Đường dẫn tệp Excel trên Google Drive (Cần thay đổi đúng đường dẫn thực tế)
file_path = '/content/drive/My Drive/Technical_Test/Database - Q3_2020.xlsx'

# Đọc file Excel
xls = pd.ExcelFile(file_path)



# Đọc từng sheet
df1 = pd.read_excel(xls, sheet_name=xls.sheet_names[0])
df2 = pd.read_excel(xls, sheet_name=xls.sheet_names[1])
df3 = pd.read_excel(xls, sheet_name=xls.sheet_names[2])




Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# I.MKT

# Ta tiến hành show bảng

In [ ]:
df1.head()

Date Channel    MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB    DucVu        CVS                  0           0   
1 2020-07-01      FB    DucVu       Mess                  0           0   
2 2020-07-01      FB    DucVu      Brand             259352       61673   
3 2020-07-01      FB  NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB  NgocNgo      Brand                  0           0   

     Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0  ...        1   1.00000  4.990000e+05   
1      0.0      0      0    0  ...        3   1.00000  4.990000e+05   
2  59968.0    271      0    0  ...        0   0.00000  0.000000e+00   
3  76992.0    386      4   26  ...       40  31.73913  1.583783e+07   
4      0.0      0      0    0  ...        0   0.00000  0.000000e+00   

   Paid Revenue 1   Giá Lead  Đơn/Lead           CPM           CPC  \
0          499000       0.00  1.000000      0.000000      0.000000   
1          499000       0.00  0.333333      0.000000      0.000000   
2               0       0.00  0.000000   4205.276215    957.018450   
3        12974000  113543.65  0.793478  47330.561287  11766.181347   
4               0       0.00  0.000000      0.000000      0.000000   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
1                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00  

[5 rows x 21 columns]

# Show số dòng và cột

In [ ]:
df1.shape

(1511, 21)

# show kiểu dữ liệu

In [ ]:
df1.dtypes

Date                              datetime64[ns]
Channel                                   object
MKTer                                     object
Chiến dịch                                object
Chi phí Marketing                          int64
Impression                                 int64
Reach                                    float64
Click                                      int64
Share                                      int64
Cmt                                        int64
Inbox                                     object
Lead MKT                                   int64
Đơn hàng                                 float64
Doanh thu                                float64
Paid Revenue 1                             int64
Giá Lead                                 float64
Đơn/Lead                                 float64
CPM                                      float64
CPC                                      float64
Giá Mess\n(Cmt + Inbox)                  float64
ME1 (ME đã bao gồm thuế + phí)           float64
dtype: object

# Tien xu li du lieu

# Kiem tra du lieu thieu

# Kiểm tra dữ liệu bị thiếu

In [ ]:
df1.isnull().sum()

Date                               0
Channel                            0
MKTer                              0
Chiến dịch                         0
Chi phí Marketing                  0
Impression                         0
Reach                              0
Click                              0
Share                              0
Cmt                                0
Inbox                              0
Lead MKT                           0
Đơn hàng                           0
Doanh thu                          0
Paid Revenue 1                     0
Giá Lead                           0
Đơn/Lead                           0
CPM                                0
CPC                                0
Giá Mess\n(Cmt + Inbox)            0
ME1 (ME đã bao gồm thuế + phí)    14
dtype: int64

# Xóa các dự liệu thiếu ở ME1

In [ ]:
df1.dropna(inplace=True)

# Show lại bảng( ở file trươc của cty em thấy nó có cả dữ liệu thiếu ở phần Channel, chiến dịch , chí phí marketing) nhưng mà do code cuad colab bị lỗi hiển thị nên không show được code nên em xin trình bày cách xử lí ở đây

*   Đối với Channel , chiến dịch ( sử dụng kĩ thuật học máy , ở đây em kiến nghị dùng decesion Tress Classfication bởi vì nó không tốn quá nhiều bộ nhớ nhưng vẫn dự đoán chính xac với tỉ lệ rất cao các nhãn và có tốc độ chay nhanh va hiệu quả)

*   Mục chi phí marketing thì khi em kiểm tra thiếu bằng lệnh unique thì thấy các dự liệu bị NaN do có Channel là Fb ma fb thì khi em check ở exel thì hàu hết là sẽ không mất chi phí nên em dùng lệnh FillNa cho NaN ở những dữ liệu thiếu là 0



# Ta sẽ hoàn toàn xư lí xong các dữ liệu thiếu

In [ ]:
df1.isnull().sum()

Date                              0
Channel                           0
MKTer                             0
Chiến dịch                        0
Chi phí Marketing                 0
Impression                        0
Reach                             0
Click                             0
Share                             0
Cmt                               0
Inbox                             0
Lead MKT                          0
Đơn hàng                          0
Doanh thu                         0
Paid Revenue 1                    0
Giá Lead                          0
Đơn/Lead                          0
CPM                               0
CPC                               0
Giá Mess\n(Cmt + Inbox)           0
ME1 (ME đã bao gồm thuế + phí)    0
dtype: int64

# Tiếp theo ta tiến hành kiểm tra xem có dữ liệu bi trùng lặp không

In [ ]:
duplicate_rows = df1.duplicated()
df1[duplicate_rows]

Date Channel  MKTer Chiến dịch  Chi phí Marketing  Impression  \
636 2020-08-05      FB  DucVu       Mess                  0           0   

     Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng  Doanh thu  \
636    0.0      0      0    0  ...        0       0.0        0.0   

     Paid Revenue 1  Giá Lead  Đơn/Lead  CPM  CPC  Giá Mess\n(Cmt + Inbox)  \
636               0       0.0       0.0  0.0  0.0                      0.0   

     ME1 (ME đã bao gồm thuế + phí)  
636                             0.0  

[1 rows x 21 columns]

In [ ]:
df1.drop_duplicates(inplace=True)
duplicate_rows = df1.duplicated()
df1[duplicate_rows]

Empty DataFrame
Columns: [Date, Channel, MKTer, Chiến dịch, Chi phí Marketing, Impression, Reach, Click, Share, Cmt, Inbox, Lead MKT, Đơn hàng, Doanh thu, Paid Revenue 1, Giá Lead, Đơn/Lead, CPM, CPC, Giá Mess
(Cmt + Inbox), ME1 (ME đã bao gồm thuế + phí)]
Index: []

[0 rows x 21 columns]

# Kiểm tra trùng lặp theo từng cột

In [ ]:
duplicates_subset = df1.duplicated(subset=['Date', 'Channel', 'MKTer', 'Đơn hàng'])
df1[duplicates_subset]

Date Channel     MKTer     Chiến dịch  Chi phí Marketing  \
1    2020-07-01      FB     DucVu           Mess                  0   
16   2020-07-01  Google  TienTran     Guu-Search              22323   
17   2020-07-01  Google  TienTran  Guu-Discovery              45845   
21   2020-07-02      FB     DucVu          Brand             306167   
34   2020-07-02  Google  TienTran         PerMax             132394   
...         ...     ...       ...            ...                ...   
1494 2020-09-29  Google  TienTran     Guu-PerMax             359239   
1498 2020-09-30      FB     DucVu          Brand             501269   
1506 2020-09-30  Google  TienTran         PerMax             441132   
1507 2020-09-30  Google  TienTran     Guu-Search               4869   
1509 2020-09-30  Google  TienTran     Guu-PerMax             129022   

      Impression    Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng  \
1              0      0.0      0      0    0  ...        3       1.0   
16           343      0.0      9      0    0  ...        0       0.0   
17          3518      0.0    148      0    0  ...        0       0.0   
21         73969  73074.0    385      0    2  ...        0       0.0   
34          5641      0.0     57      0    0  ...        1       1.0   
...          ...      ...    ...    ...  ...  ...      ...       ...   
1494       16572      0.0    257      0    0  ...        0       0.0   
1498           0      0.0      0      0    0  ...        0       0.0   
1506       16658      0.0    116      0    0  ...        0       0.0   
1507         172      0.0      5      0    0  ...        0       0.0   
1509        2986      0.0     22      0    0  ...        1       0.0   

      Doanh thu  Paid Revenue 1  Giá Lead  Đơn/Lead           CPM  \
1      499000.0          499000       0.0  0.333333      0.000000   
16          0.0               0       0.0  0.000000  65081.632653   
17          0.0               0       0.0  0.000000  13031.552018   
21          0.0               0       0.0  0.000000   4139.125850   
34     499000.0          499000  132394.0  1.000000  23469.952136   
...         ...             ...       ...       ...           ...   
1494        0.0               0       0.0  0.000000  21677.468018   
1498        0.0               0       0.0  0.000000      0.000000   
1506        0.0               0       0.0  0.000000  26481.690479   
1507        0.0               0       0.0  0.000000  28308.139535   
1509        0.0               0  129022.0  0.000000  43208.975218   

              CPC  Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
1        0.000000                 0.000000                            0.00  
16    2480.333333                 0.000000                        22323.00  
17     309.763514                 0.000000                        45845.00  
21     795.238961            102055.666667                       330660.36  
34    2322.701754                 0.000000                       132394.00  
...           ...                      ...                             ...  
1494  1397.817121                 0.000000                       359239.00  
1498     0.000000                 0.000000                       541370.52  
1506  3802.862069                 0.000000                       441132.00  
1507   973.800000                 0.000000                         4869.00  
1509  5864.636364                 0.000000                       129022.00  

[337 rows x 21 columns]

# Ta thấy có các dòng bị trùng lặp là 337 dòng

In [ ]:
duplicates_count = df1.duplicated(subset=['Date', 'Channel', 'MKTer', 'Đơn hàng']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 337


# Ta tiến hành xóa trùng lặp

In [ ]:
df1 = df1.drop_duplicates(subset=['Date', 'Channel', 'MKTer', 'Đơn hàng'], keep='first')


# Sau khi xóa các dữ liệu bị trùng lặp (chỉ xóa trùng lặp ở phía sau , không xóa những dòng gốc)

In [ ]:
duplicates_count = df1.duplicated(subset=['Date', 'Channel', 'MKTer', 'Đơn hàng']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 0


In [ ]:
df1.head(5)

Date Channel      MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB      DucVu        CVS                  0           0   
2 2020-07-01      FB      DucVu      Brand             259352       61673   
3 2020-07-01      FB    NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB    NgocNgo      Brand                  0           0   
5 2020-07-01      FB  GiangPhan        CVS             488991        4596   

     Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0  ...        1   1.00000  4.990000e+05   
2  59968.0    271      0    0  ...        0   0.00000  0.000000e+00   
3  76992.0    386      4   26  ...       40  31.73913  1.583783e+07   
4      0.0      0      0    0  ...        0   0.00000  0.000000e+00   
5   4389.0     19      1    3  ...        1   1.00000  4.990000e+05   

   Paid Revenue 1   Giá Lead  Đơn/Lead            CPM           CPC  \
0          499000       0.00  1.000000       0.000000      0.000000   
2               0       0.00  0.000000    4205.276215    957.018450   
3        12974000  113543.65  0.793478   47330.561287  11766.181347   
4               0       0.00  0.000000       0.000000      0.000000   
5          499000  488991.00  1.000000  106394.908616  25736.368421   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00  
5            162997.000000                       528110.28  

[5 rows x 21 columns]

In [ ]:
df1.shape

(1159, 21)

In [ ]:
df1.head()

Date Channel      MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB      DucVu        CVS                  0           0   
2 2020-07-01      FB      DucVu      Brand             259352       61673   
3 2020-07-01      FB    NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB    NgocNgo      Brand                  0           0   
5 2020-07-01      FB  GiangPhan        CVS             488991        4596   

     Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0  ...        1   1.00000  4.990000e+05   
2  59968.0    271      0    0  ...        0   0.00000  0.000000e+00   
3  76992.0    386      4   26  ...       40  31.73913  1.583783e+07   
4      0.0      0      0    0  ...        0   0.00000  0.000000e+00   
5   4389.0     19      1    3  ...        1   1.00000  4.990000e+05   

   Paid Revenue 1   Giá Lead  Đơn/Lead            CPM           CPC  \
0          499000       0.00  1.000000       0.000000      0.000000   
2               0       0.00  0.000000    4205.276215    957.018450   
3        12974000  113543.65  0.793478   47330.561287  11766.181347   
4               0       0.00  0.000000       0.000000      0.000000   
5          499000  488991.00  1.000000  106394.908616  25736.368421   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00  
5            162997.000000                       528110.28  

[5 rows x 21 columns]

In [ ]:
df1.dtypes

Date                              datetime64[ns]
Channel                                   object
MKTer                                     object
Chiến dịch                                object
Chi phí Marketing                          int64
Impression                                 int64
Reach                                    float64
Click                                      int64
Share                                      int64
Cmt                                        int64
Inbox                                     object
Lead MKT                                   int64
Đơn hàng                                 float64
Doanh thu                                float64
Paid Revenue 1                             int64
Giá Lead                                 float64
Đơn/Lead                                 float64
CPM                                      float64
CPC                                      float64
Giá Mess\n(Cmt + Inbox)                  float64
ME1 (ME đã bao gồm thuế + phí)           float64
dtype: object

# Dong nhat du lieu

# Kiểm tra cột kiểu object xem có lỗi định dạng không(lệnh unique)

In [ ]:

print(df1['Inbox'].unique())



[0 1 18 '-']


In [ ]:
 df1['Inbox'] = pd.to_numeric(df1['Inbox'], errors='coerce')


In [ ]:
print(df1['Channel'].unique())

['FB' 'Google' 'Tiktok']


In [ ]:
print(df1['Chiến dịch'].unique())

['CVS' 'Brand' 'Search' 'Discovery' 'PerMax' 'Guu-PerMax' 'Mess'
 'Guu-Discovery' 'Guu-Search']


# Ở côt MKTer ta thấy có rất nhiều nhân viên ghi số ở sau , điều này là không cần thiết bời mỗi nhân viên đều mã số riêng và tên nên đúng định dạng chỉ chữ

In [ ]:
print(df1['MKTer'].unique())

['DucVu' 'NgocNgo' 'GiangPhan' 'AnhNguyen08' 'TungTran' 'AnhTrinh'
 'AnhNgo01' 'YenNguyen02' 'ThuHoang' 'HuyDo' 'TienTran' 'NhiLanh'
 'HuyenNguyen05' 'TrangDang' 'TuanTran' 'LinhHoang' 'LyLuu']


# Ta tiến hành xóa cách và só

In [ ]:
df1['MKTer'] = df1['MKTer'].str.replace(r'\d+$', '', regex=True)


# Kiểm tra lại dữ liệu sau xóa

In [ ]:
print(df1['MKTer'].unique())


['DucVu' 'NgocNgo' 'GiangPhan' 'AnhNguyen' 'TungTran' 'AnhTrinh' 'AnhNgo'
 'YenNguyen' 'ThuHoang' 'HuyDo' 'TienTran' 'NhiLanh' 'HuyenNguyen'
 'TrangDang' 'TuanTran' 'LinhHoang' 'LyLuu']


In [ ]:
df1.head(5)

Date Channel      MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB      DucVu        CVS                  0           0   
2 2020-07-01      FB      DucVu      Brand             259352       61673   
3 2020-07-01      FB    NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB    NgocNgo      Brand                  0           0   
5 2020-07-01      FB  GiangPhan        CVS             488991        4596   

     Reach  Click  Share  Cmt  ...  Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0  ...         1   1.00000  4.990000e+05   
2  59968.0    271      0    0  ...         0   0.00000  0.000000e+00   
3  76992.0    386      4   26  ...        40  31.73913  1.583783e+07   
4      0.0      0      0    0  ...         0   0.00000  0.000000e+00   
5   4389.0     19      1    3  ...         1   1.00000  4.990000e+05   

   Paid Revenue 1   Giá Lead  Đơn/Lead            CPM           CPC  \
0          499000       0.00  1.000000       0.000000      0.000000   
2               0       0.00  0.000000    4205.276215    957.018450   
3        12974000  113543.65  0.793478   47330.561287  11766.181347   
4               0       0.00  0.000000       0.000000      0.000000   
5          499000  488991.00  1.000000  106394.908616  25736.368421   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00  
5            162997.000000                       528110.28  

[5 rows x 21 columns]

# II.Sales

# Tien xu li du lieu va dong nhat du lieu

# Tiến hành show bảng df2

In [ ]:
df2.head(5)

Unnamed: 0                 Giờ        Khách hàng          SĐT Channel  \
0   1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1   1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2   1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3   1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4   1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content   Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen08      Dathang   
1        CVS         cay      NgocNgo      Dathang   
2        CVS      matcon      NgocNgo      Dathang   
3        CVS         cay      NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen02      Dathang   

  Sales Admin xác nhận Type of Lead  ... Số lần tương tác   Ngày gọi  \
0                          DATTHANG  ...              1.0 2020-07-01   
1                          DATTHANG  ...              1.0 2020-07-01   
2                          DATTHANG  ...              1.0 2020-07-03   
3                          DATTHANG  ...              1.0 2020-07-03   
4                          DATTHANG  ...              1.0 2020-07-03   

             Trạng thái                               Level Ngày hẹn gọi lại  \
0  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
1              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
2  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
3  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
4              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   

  Close Date     Tỉnh/TP Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0        NaT         NaN              1.0          140000.0     499000  
1 2023-07-04  Quảng Ninh              1.0          140000.0     499000  
2        NaT         NaN              1.0          140000.0     499000  
3        NaT         NaN              1.0          140000.0     499000  
4 2023-07-05      Hà Nội              1.0          140000.0     499000  

[5 rows x 21 columns]

# Show số dòng và số cột

In [ ]:
df2.shape

(9675, 21)

# Show kiểu dữ liệu

In [ ]:
df2.dtypes

Unnamed: 0                                   object
Giờ                                  datetime64[ns]
Khách hàng                                   object
SĐT                                         float64
Channel                                      object
Chiến dịch                                   object
Content                                      object
Marketer 2                                   object
Type of Lead                                 object
Sales Admin xác nhận Type of Lead            object
Sales                                        object
Số lần tương tác                            float64
Ngày gọi                             datetime64[ns]
Trạng thái                                   object
Level                                        object
Ngày hẹn gọi lại                     datetime64[ns]
Close Date                           datetime64[ns]
Tỉnh/TP                                      object
Số lượng bộ sách                            float64
Số tiền giảm giá                            float64
Tổng tiền                                     int64
dtype: object

# Xu li du lieu thieu

# Kiểm tra xem có các dữ liệu thiếu nào

In [ ]:
df2.isnull().sum()

Unnamed: 0                            171
Giờ                                   171
Khách hàng                            171
SĐT                                   171
Channel                               171
Chiến dịch                            171
Content                               175
Marketer 2                           2894
Type of Lead                          177
Sales Admin xác nhận Type of Lead    4227
Sales                                 171
Số lần tương tác                      179
Ngày gọi                             9471
Trạng thái                            357
Level                                1691
Ngày hẹn gọi lại                     9627
Close Date                           3514
Tỉnh/TP                              6281
Số lượng bộ sách                     2624
Số tiền giảm giá                     2633
Tổng tiền                               0
dtype: int64

In [ ]:
df2.dtypes

Unnamed: 0                                   object
Giờ                                  datetime64[ns]
Khách hàng                                   object
SĐT                                         float64
Channel                                      object
Chiến dịch                                   object
Content                                      object
Marketer 2                                   object
Type of Lead                                 object
Sales Admin xác nhận Type of Lead            object
Sales                                        object
Số lần tương tác                            float64
Ngày gọi                             datetime64[ns]
Trạng thái                                   object
Level                                        object
Ngày hẹn gọi lại                     datetime64[ns]
Close Date                           datetime64[ns]
Tỉnh/TP                                      object
Số lượng bộ sách                            float64
Số tiền giảm giá                            float64
Tổng tiền                                     int64
dtype: object

# Ta tiến hành kiểm tỷ lệ phần trăm giá trị thiếu (Nếu dữ liệu ở cột bí thiếu quá 60% thì dữ liệu này e khuyến nghị loại bỏ cột đó bởi vì tính ứng dụng thấp) còn nếu dữ liệu thiếu dưới 60% thì chúng ta sẽ tiến hành các phương pháp để điền

# Ví dụ ở cột giờ này mặc dù tỉ lệ dữ liệu thiếu chỉ ở mức 1,77% nhưng mà dữ liệu thiếu cũng rất ít ( xóa đi cũng không ảnh hưởng đến quá trình phân tích và dự báo )nên em khuyến nghị xóa bỏ những dòng thiếu

In [ ]:


def check_missing_values(df2, column_name):
    total_missing = df2[column_name].isna().sum()
    total_rows = len(df2)
    missing_percentage = (total_missing / total_rows) * 100

    print(f"Số lượng giá trị thiếu trong cột '{column_name}': {total_missing}")
    print(f"Tỷ lệ phần trăm giá trị thiếu: {missing_percentage:.2f}%")

check_missing_values(df2, 'Giờ')


Số lượng giá trị thiếu trong cột 'Giờ': 171
Tỷ lệ phần trăm giá trị thiếu: 1.77%


# Tiến hành tương tụ với nhưng cột như giờ , khách hàng,SDT,...

In [ ]:
df2 = df2.dropna(subset=['Unnamed: 0','Giờ','Khách hàng','SĐT','Channel','Chiến dịch','Số lần tương tác'])

# Kiểm tra dữ liệu bị sau khi xử lí

In [ ]:
df2.isnull().sum()

Unnamed: 0                              0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 4
Marketer 2                           2719
Type of Lead                            6
Sales Admin xác nhận Type of Lead    4048
Sales                                   0
Số lần tương tác                        0
Ngày gọi                             9292
Trạng thái                            180
Level                                1513
Ngày hẹn gọi lại                     9448
Close Date                           3336
Tỉnh/TP                              6102
Số lượng bộ sách                     2446
Số tiền giảm giá                     2455
Tổng tiền                               0
dtype: int64

# Ta thấy ở content có 4 dự liệu thiếu, ta thử tiến hành phân tích xem nó bị thiếu như nào( lệnh isna)

In [ ]:
missing_Content_rows = df2[df2['Content'].isna()]

# Ta thấy content chư yếu thiếu ở bên Channel FB

In [ ]:
missing_Content_rows

Unnamed: 0                 Giờ  Khách hàng          SĐT Channel  \
4228  23-9-2020 2023-09-23 10:43:08       Hường  123404228.0      FB   
6004  29-7-2020 2023-07-29 12:20:17   Nguyệt Vũ  123406002.0      FB   
8175   6-8-2020 2023-08-06 13:39:25  Thùy Khiêm  123408173.0      FB   
9114   9-7-2020 2023-07-09 06:22:28     Bui Nga  123409112.0      FB   

     Chiến dịch Content   Marketer 2 Type of Lead  \
4228        CVS     NaN          NaN      Dathang   
6004        CVS     NaN  YenNguyen02      Dathang   
8175        CVS     NaN  YenNguyen02        Tuvan   
9114        CVS     NaN    GiangPhan      Dathang   

     Sales Admin xác nhận Type of Lead  ... Số lần tương tác  Ngày gọi  \
4228                               NaN  ...              2.0       NaT   
6004                          DATTHANG  ...              1.0       NaT   
8175                               NaN  ...              5.0       NaT   
9114                          DATTHANG  ...              1.0       NaT   

     Trạng thái                               Level Ngày hẹn gọi lại  \
4228   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
6004   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
8175   Đặt hàng             L7.1 - Chờ chuyển khoản              NaT   
9114   Đặt hàng                                 NaN              NaT   

     Close Date     Tỉnh/TP Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
4228 2023-09-27         NaN              1.0          140000.0     499000  
6004 2023-08-01    Bắc Ninh              1.0          140000.0     499000  
8175        NaT  Bình Dương              1.0          240000.0     399000  
9114 2023-07-11      Hà Nội              1.0          140000.0     499000  

[4 rows x 21 columns]

# Ta tiến hành điển các dữ liệu phổ biến ở cotent nhưng với điều kiện là Channel phải là FB

In [ ]:
most_frequent_fb_content = df2.loc[df2["Channel"] == "FB", "Content"].mode().iat[0] if not df2.loc[df2["Channel"] == "FB", "Content"].mode().empty else None


df2.loc[(df2["Channel"] == "FB") & (df2["Content"].isna()), "Content"] = most_frequent_fb_content


# Thử kiểm tra 1 cột bị thiếu vừa xong , ta thấy nó đã được fill đầy bởi 1 cotent phổ biến cùa FB

In [ ]:
df2.iloc[4228]

Unnamed: 0                                                    23-9-2020
Giờ                                                 2023-09-23 11:15:32
Khách hàng                                             NGuyễn Hồng Minh
SĐT                                                         123404233.0
Channel                                                              FB
Chiến dịch                                                          CVS
Content                                                   ducvu_benhlau
Marketer 2                                                          NaN
Type of Lead                                                    Dathang
Sales Admin xác nhận Type of Lead                              DATTHANG
Sales                                                          ThuyTran
Số lần tương tác                                                    1.0
Ngày gọi                                                            NaT
Trạng thái                                                     Đặt hàng
Level                                L8.2 - Khách đã thanh toán toàn bộ
Ngày hẹn gọi lại                                                    NaT
Close Date                                          2023-09-27 00:00:00
Tỉnh/TP                                                             NaN
Số lượng bộ sách                                                    1.0
Số tiền giảm giá                                               140000.0
Tổng tiền                                                        499000
Name: 4233, dtype: object

In [ ]:
df2.isnull().sum()

Unnamed: 0                              0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 0
Marketer 2                           2719
Type of Lead                            6
Sales Admin xác nhận Type of Lead    4048
Sales                                   0
Số lần tương tác                        0
Ngày gọi                             9292
Trạng thái                            180
Level                                1513
Ngày hẹn gọi lại                     9448
Close Date                           3336
Tỉnh/TP                              6102
Số lượng bộ sách                     2446
Số tiền giảm giá                     2455
Tổng tiền                               0
dtype: int64

# Tiếp theo ta xử lí cột Type of Lead

In [ ]:
missing_Type_of_Lead_rows = df2[df2['Type of Lead'].isna()]

# Ta tiến hành radom 1 lựa chọn trong Type of Lead

In [ ]:

existing_values = df2['Type of Lead'].dropna().unique()


df2.loc[df2["Type of Lead"].isna(), "Type of Lead"] = np.random.choice(existing_values, size=df2["Type of Lead"].isna().sum())

# Tiếp theo là cột Marketer 2, ta tiến hành xem các giá trị trong cột của nó

In [ ]:
print(df2['Marketer 2'].unique())

['AnhNguyen08' 'NgocNgo' 'YenNguyen02' 'TungTran' 'GiangPhan' 'AnhNgo01'
 'TrangDang' 'TienTran' 'DucVu' 'TuanTran' 'NhiLanh' 'ThuHoang'
 'LinhHoang' nan 'AnhTrinh' 'HuyDo' 'ht.content.05' 'LyLuu'
 'HuyenNguyen05' '   ']


# Nhận thấy có nhiều định dạng sai(chữ lẫn số, nhiều dấu tab) ta tiến hành xóa bỏ bành lệnh .replace

In [ ]:
df2['Marketer 2'] = df2['Marketer 2'].str.replace(r'\d+$', '   ', regex=True)

# Sau khi xóa bỏ

In [ ]:
print(df2['Marketer 2'].unique())

['AnhNguyen   ' 'NgocNgo' 'YenNguyen   ' 'TungTran' 'GiangPhan'
 'AnhNgo   ' 'TrangDang' 'TienTran' 'DucVu' 'TuanTran' 'NhiLanh'
 'ThuHoang' 'LinhHoang' nan 'AnhTrinh' 'HuyDo' 'ht.content.   ' 'LyLuu'
 'HuyenNguyen   ' '   ']


ở phần này do dữ liệu thiếu là rất nhiều nên em khuyến nghịc dùng học máy( ở đây em dùng desecsion tree classfication) các bước chi tiết e có note ở code.

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Marketer 2']  # Các cột cần dự đoán
feature_cols = ["Channel", "Chiến dịch", "Content", "Khách hàng"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df2[col] = le.fit_transform(df2[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần
    df_cleaned_train = df2[df2[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df2[df2[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df2 với giá trị đã điền
    df2.loc[df2[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df2[columns_missing].isnull().sum())  # Kiểm tra còn NaN không


Đang xử lý cột: Marketer 2
Số giá trị NaN sau khi xử lý:
Marketer 2    0
dtype: int64


<ipython-input-36-4436b2baf65e>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


Mã hóa lại cột về dạng chữ


In [ ]:
for col in feature_cols:
    if col in encoders:  # Kiểm tra xem cột có trong danh sách mã hóa không
        df2[col] = encoders[col].inverse_transform(df2[col])


# Kết quả

In [ ]:
df2.head(5)

Unnamed: 0                 Giờ        Khách hàng          SĐT Channel  \
0   1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1   1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2   1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3   1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4   1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content    Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen         Dathang   
1        CVS         cay       NgocNgo      Dathang   
2        CVS      matcon       NgocNgo      Dathang   
3        CVS         cay       NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen         Dathang   

  Sales Admin xác nhận Type of Lead  ... Số lần tương tác   Ngày gọi  \
0                          DATTHANG  ...              1.0 2020-07-01   
1                          DATTHANG  ...              1.0 2020-07-01   
2                          DATTHANG  ...              1.0 2020-07-03   
3                          DATTHANG  ...              1.0 2020-07-03   
4                          DATTHANG  ...              1.0 2020-07-03   

             Trạng thái                               Level Ngày hẹn gọi lại  \
0  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
1              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
2  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
3  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
4              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   

  Close Date     Tỉnh/TP Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0        NaT         NaN              1.0          140000.0     499000  
1 2023-07-04  Quảng Ninh              1.0          140000.0     499000  
2        NaT         NaN              1.0          140000.0     499000  
3        NaT         NaN              1.0          140000.0     499000  
4 2023-07-05      Hà Nội              1.0          140000.0     499000  

[5 rows x 21 columns]

In [ ]:
df2.dtypes

Unnamed: 0                                   object
Giờ                                  datetime64[ns]
Khách hàng                                   object
SĐT                                         float64
Channel                                      object
Chiến dịch                                   object
Content                                      object
Marketer 2                                   object
Type of Lead                                 object
Sales Admin xác nhận Type of Lead            object
Sales                                        object
Số lần tương tác                            float64
Ngày gọi                             datetime64[ns]
Trạng thái                                   object
Level                                        object
Ngày hẹn gọi lại                     datetime64[ns]
Close Date                           datetime64[ns]
Tỉnh/TP                                      object
Số lượng bộ sách                            float64
Số tiền giảm giá                            float64
Tổng tiền                                     int64
dtype: object

In [ ]:
df2.isnull().sum()

Unnamed: 0                              0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 0
Marketer 2                              0
Type of Lead                            0
Sales Admin xác nhận Type of Lead    4048
Sales                                   0
Số lần tương tác                        0
Ngày gọi                             9292
Trạng thái                            180
Level                                1513
Ngày hẹn gọi lại                     9448
Close Date                           3336
Tỉnh/TP                              6102
Số lượng bộ sách                     2446
Số tiền giảm giá                     2455
Tổng tiền                               0
dtype: int64

# Ví dụ ở cột này dữ liệu thiếu ở cột lên đến 97,85% , dữ liệu thiếu là quá lớn( có thể là lỗi do ng ghi hoặc do cột này thực sự chưa quá cần thiết), Tương tự với cột ngày gọi lại

In [ ]:
def check_missing_values(df2, column_name):
    total_missing = df2[column_name].isna().sum()  # Số lượng giá trị thiếu
    total_rows = len(df2)  # Tổng số dòng
    missing_percentage = (total_missing / total_rows) * 100  # Phần trăm dữ liệu thiếu

    print(f"Số lượng giá trị thiếu trong cột '{column_name}': {total_missing}")
    print(f"Tỷ lệ phần trăm giá trị thiếu: {missing_percentage:.2f}%")

# Ví dụ: kiểm tra dữ liệu thiếu trong cột 'subcategory2'
check_missing_values(df2, 'Ngày gọi')


Số lượng giá trị thiếu trong cột 'Ngày gọi': 9292
Tỷ lệ phần trăm giá trị thiếu: 97.85%


In [ ]:
df2.isnull().sum()

Unnamed: 0                              0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 0
Marketer 2                              0
Type of Lead                            0
Sales Admin xác nhận Type of Lead    4048
Sales                                   0
Số lần tương tác                        0
Ngày gọi                             9292
Trạng thái                            180
Level                                1513
Ngày hẹn gọi lại                     9448
Close Date                           3336
Tỉnh/TP                              6102
Số lượng bộ sách                     2446
Số tiền giảm giá                     2455
Tổng tiền                               0
dtype: int64

Ta tiến hành xóa bỏ

# Ta tiến hành xóa 2 cột

In [ ]:
df2.drop(columns=["Ngày gọi", "Ngày hẹn gọi lại"], inplace=True)


# Tiếp theo ở cột Sales Admin xác nhận Type of Lead, ta tiến hành kiểm tra các giá trị thiếu ở cột

In [ ]:
print(df2['Sales Admin xác nhận Type of Lead'].unique())

['DATTHANG' nan]


# Ta thấy ở đây chỉ có  giá trị la DATTHANG , vâyh ta tiến hành điền dữ liệu bị thiếu bằng Fillna

In [ ]:
df2["Sales Admin xác nhận Type of Lead"].fillna("DATTHANG", inplace=True)


# Tiếp theo ở cột Sales Admin xác nhận Type of Lead, ta tiến hành kiểm tra các giá trị thiếu ở cột

In [ ]:
print(df2['Trạng thái'].unique())

['Hủy đơn sau khi chốt' 'Đặt hàng' 'Suy nghĩ thêm/Trả lời sau'
 'Không nghe máy' 'Từ chối' 'Hẹn lúc khác gọi lại' 'Sai số' 'Trùng'
 'Thuê bao' 'Trêu đùa' nan]


# Tiếp tục dùng phương pháp học máy để dự đoán các dữ liệu thiếu ở cột trạng thái

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Trạng thái']  # Các cột cần dự đoán
feature_cols = ["Channel", "Chiến dịch", "Content", "Khách hàng","Marketer 2"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df2[col] = le.fit_transform(df2[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần
    df_cleaned_train = df2[df2[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df2[df2[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df2 với giá trị đã điền
    df2.loc[df2[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df2[columns_missing].isnull().sum())  # Kiểm tra còn NaN không


Đang xử lý cột: Trạng thái
Số giá trị NaN sau khi xử lý:
Trạng thái    0
dtype: int64


<ipython-input-45-c1ec4814d755>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


In [ ]:
for col in feature_cols:
    if col in encoders:  # Kiểm tra xem cột có trong danh sách mã hóa không
        df2[col] = encoders[col].inverse_transform(df2[col])


# Ta tiến hành kiểm tra giá trị ở cột level

In [ ]:
print(df2['Level'].unique())

['L7.2 - Hủy đơn sau khi chốt' 'L8.2 - Khách đã thanh toán toàn bộ'
 'L3.1 - SQL - Đã được tư vấn đầy đủ' nan 'L1.1 - Không nghe máy/Thuê bao'
 'L5.2 - Từ chối' 'L4.2 - Băn khoăn chất lượng SP'
 'L2.1 - MQL - Hẹn lúc khác gọi lại' 'L1.4 - Trêu đùa'
 'L7.1 - Chờ chuyển khoản' 'L4.1 - Băn khoăn về giá'
 'L3.2 - SQL - Chưa được tư vấn đầy đủ' 'L4.3 - Suy nghĩ thêm/Trả lời sau'
 'L2.2 - MQL - Không còn nhu cầu' 'L1.3 - Không đăng ký'
 'L5.1 - Hỏi ý kiến người thân' 'L3.3 - SQL - Chủ động dừng tư vấn'
 'C3.1 - Sai số' 'L8.1 - Khách đã thanh toán một phần' 'C3.2 - Trùng']


# Ta nhận thấy những cái nhứ L7.2 -Hủy đơn... là rất dai và gây nhầm lẫn. Khuyến nghị cuae em là nên tạo 1 bangr gồm 2 cột trong đó cột trái là kí hiệu cột phải là giải thích nó để làm gì

# Ta tiến hành xóa những cái từ - trở đi

In [ ]:
df2['Level'] = df2['Level'].str.replace(r'\s*-\s*.*', '', regex=True)


# Kiểm tra lại cột sau khi xử lí

In [ ]:
print(df2['Level'].unique())

['L7.2' 'L8.2' 'L3.1' nan 'L1.1' 'L5.2' 'L4.2' 'L2.1' 'L1.4' 'L7.1' 'L4.1'
 'L3.2' 'L4.3' 'L2.2' 'L1.3' 'L5.1' 'L3.3' 'C3.1' 'L8.1' 'C3.2']


# Ta lại tiến hành dự đoán dữ liệu bằng phương pháp học máy

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Level']  # Các cột cần dự đoán
feature_cols = ["Channel", "Chiến dịch", "Content", "Khách hàng","Marketer 2","Trạng thái"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df2[col] = le.fit_transform(df2[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần
    df_cleaned_train = df2[df2[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df2[df2[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df2 với giá trị đã điền
    df2.loc[df2[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df2[columns_missing].isnull().sum())  # Kiểm tra còn NaN không


Đang xử lý cột: Level
Số giá trị NaN sau khi xử lý:
Level    0
dtype: int64


<ipython-input-50-28419dd0b5bf>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


In [ ]:
for col in feature_cols:
    if col in encoders:  # Kiểm tra xem cột có trong danh sách mã hóa không
        df2[col] = encoders[col].inverse_transform(df2[col])


# Ta đổi lại cột tên là Unamed : 0 thành date để sau này phân tích dễ dáng hơn

In [ ]:
df2.rename(columns={"Unnamed: 0": "date"}, inplace=True)

# Ta xem lại bảng

In [ ]:
df2.head(5)

date                 Giờ        Khách hàng          SĐT Channel  \
0  1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1  1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2  1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3  1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4  1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content    Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen         Dathang   
1        CVS         cay       NgocNgo      Dathang   
2        CVS      matcon       NgocNgo      Dathang   
3        CVS         cay       NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen         Dathang   

  Sales Admin xác nhận Type of Lead     Sales  Số lần tương tác  \
0                          DATTHANG  ThuyTran               1.0   
1                          DATTHANG  ThuyTran               1.0   
2                          DATTHANG  ThuyTran               1.0   
3                          DATTHANG  ThuyTran               1.0   
4                          DATTHANG  ThuyTran               1.0   

             Trạng thái Level Close Date     Tỉnh/TP  Số lượng bộ sách  \
0  Hủy đơn sau khi chốt  L7.2        NaT         NaN               1.0   
1              Đặt hàng  L8.2 2023-07-04  Quảng Ninh               1.0   
2  Hủy đơn sau khi chốt  L7.2        NaT         NaN               1.0   
3  Hủy đơn sau khi chốt  L7.2        NaT         NaN               1.0   
4              Đặt hàng  L8.2 2023-07-05      Hà Nội               1.0   

   Số tiền giảm giá  Tổng tiền  
0          140000.0     499000  
1          140000.0     499000  
2          140000.0     499000  
3          140000.0     499000  
4          140000.0     499000

# Ta nhận thấy phần lớn các cột Close Date bị thiếu là do những đơn hàng đó thường bị hủy hoặcc chỉ cân nhắc nên không điền nên em khuyến nghị nên điền những cột đó có Close Date là 0

In [ ]:
df2.loc[(df2['Trạng thái'] != 'Đặt hàng') & (df2['Close Date'].isna()), 'Close Date'] = 0


Ta tiến hành xem lại bảng sau khi điền

In [ ]:
df2.head(5)

date                 Giờ        Khách hàng          SĐT Channel  \
0  1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1  1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2  1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3  1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4  1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content    Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen         Dathang   
1        CVS         cay       NgocNgo      Dathang   
2        CVS      matcon       NgocNgo      Dathang   
3        CVS         cay       NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen         Dathang   

  Sales Admin xác nhận Type of Lead     Sales  Số lần tương tác  \
0                          DATTHANG  ThuyTran               1.0   
1                          DATTHANG  ThuyTran               1.0   
2                          DATTHANG  ThuyTran               1.0   
3                          DATTHANG  ThuyTran               1.0   
4                          DATTHANG  ThuyTran               1.0   

             Trạng thái Level           Close Date     Tỉnh/TP  \
0  Hủy đơn sau khi chốt  L7.2                    0         NaN   
1              Đặt hàng  L8.2  2023-07-04 00:00:00  Quảng Ninh   
2  Hủy đơn sau khi chốt  L7.2                    0         NaN   
3  Hủy đơn sau khi chốt  L7.2                    0         NaN   
4              Đặt hàng  L8.2  2023-07-05 00:00:00      Hà Nội   

   Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0               1.0          140000.0     499000  
1               1.0          140000.0     499000  
2               1.0          140000.0     499000  
3               1.0          140000.0     499000  
4               1.0          140000.0     499000

# Kiểm tra lại dữ liệu thiếu

In [ ]:
df2.isnull().sum()

date                                    0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 0
Marketer 2                              0
Type of Lead                            0
Sales Admin xác nhận Type of Lead       0
Sales                                   0
Số lần tương tác                        0
Trạng thái                              0
Level                                   0
Close Date                            359
Tỉnh/TP                              6102
Số lượng bộ sách                     2446
Số tiền giảm giá                     2455
Tổng tiền                               0
dtype: int64

# Thấy vẫn còn 359 dòng ở Close Date bị thiếu mà mặc dù đã đặt hàng( có thể là do hệ thống bị lỗi hoặc đơn bị trùng lắp) nên em khuyến nghị nên xóa 359 dòng này

In [ ]:
missing_rows = df2[df2["Close Date"].isna()]
missing_rows


date                 Giờ           Khách hàng          SĐT Channel  \
106    1-8-2020 2023-08-01 17:11:40  Trần Thị Hồng Châu   123400106.0      FB   
173    1-9-2020 2023-09-01 09:54:42          Hien Nguyen  123400173.0      FB   
242    1-9-2020 2023-09-01 20:55:09         Habbeld Thảo  123400242.0      FB   
334   10-7-2020 2023-07-10 10:33:14    Nguyễn Hồng Nhung  123400334.0      FB   
335   10-7-2020 2023-07-10 10:34:27       Nguyen Van Anh  123400335.0      FB   
...         ...                 ...                  ...          ...     ...   
9485   9-9-2020 2023-09-09 21:57:21     Phương Thảo Trần  123409483.0      FB   
9488   9-9-2020 2023-09-09 22:08:20      Trịnh Thị Chung  123409486.0      FB   
9492   9-9-2020 2023-09-09 19:54:12          Nguyễn Hằng  123409490.0  Google   
9500   9-9-2020 2023-09-09 15:30:20      Nguyễn Tiết Hài  123409498.0  Google   
9501   9-9-2020 2023-09-09 14:49:21       vương thị hợp   123409499.0  Google   

     Chiến dịch          Content    Marketer 2 Type of Lead  \
106         CVS  docthu_vietkieu  YenNguyen         Dathang   
173         CVS            lop11         DucVu        Tuvan   
242         CVS            lop11         DucVu        Tuvan   
334         CVS              bcs  YenNguyen           Tuvan   
335         CVS       condau_sto  YenNguyen           Tuvan   
...         ...              ...           ...          ...   
9485        CVS            lop11         DucVu      Dathang   
9488        CVS            bodoi  YenNguyen         Dathang   
9492  Discovery             2010      TienTran      Dathang   
9500     PerMax              909      TienTran      Dathang   
9501     Search             2809      TienTran      Dathang   

     Sales Admin xác nhận Type of Lead          Sales  Số lần tương tác  \
106                           DATTHANG       ThuyTran               1.0   
173                           DATTHANG          AnhHa               2.0   
242                           DATTHANG        NamPham               2.0   
334                           DATTHANG  HuyenNguyen04               2.0   
335                           DATTHANG      HuongDinh               6.0   
...                                ...            ...               ...   
9485                          DATTHANG       ThuyTran               1.0   
9488                          DATTHANG       ThuyTran               1.0   
9492                          DATTHANG       ThuyTran               1.0   
9500                          DATTHANG       ThuyTran               1.0   
9501                          DATTHANG       ThuyTran               1.0   

     Trạng thái Level Close Date    Tỉnh/TP  Số lượng bộ sách  \
106    Đặt hàng  L7.2        NaT    Bến Tre               1.0   
173    Đặt hàng  L7.1        NaT  Thanh Hóa               1.0   
242    Đặt hàng  L7.1        NaT     Hà Nội               1.0   
334    Đặt hàng  L8.2        NaT        NaN               NaN   
335    Đặt hàng  L8.2        NaT        NaN               NaN   
...         ...   ...        ...        ...               ...   
9485   Đặt hàng  L7.1        NaT        NaN               1.0   
9488   Đặt hàng  L7.1        NaT        NaN               1.0   
9492   Đặt hàng  L5.2        NaT        NaN               1.0   
9500   Đặt hàng  L5.2        NaT        NaN               1.0   
9501   Đặt hàng  L5.2        NaT        NaN               1.0   

      Số tiền giảm giá  Tổng tiền  
106           140000.0     499000  
173           140000.0     499000  
242           240000.0     399000  
334                NaN          0  
335                NaN          0  
...                ...        ...  
9485          140000.0     499000  
9488          140000.0     499000  
9492          140000.0     499000  
9500          140000.0     499000  
9501          140000.0     499000  

[359 rows x 19 columns]

In [ ]:
df2 = df2.dropna(subset=['Close Date'])

# Ta tiến hành xử lí cột Tỉnh/TP bằng phương pháp học máy

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Tỉnh/TP']  # Các cột cần dự đoán
feature_cols = ["Channel", "Chiến dịch", "Content", "Khách hàng","Marketer 2"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df2[col] = le.fit_transform(df2[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần
    df_cleaned_train = df2[df2[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df2[df2[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df2 với giá trị đã điền
    df2.loc[df2[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df2[columns_missing].isnull().sum())  # Kiểm tra còn NaN không


Đang xử lý cột: Tỉnh/TP
Số giá trị NaN sau khi xử lý:
Tỉnh/TP    0
dtype: int64


<ipython-input-59-062993d9d50f>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


In [ ]:
df2.isnull().sum()

date                                    0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 0
Marketer 2                              0
Type of Lead                            0
Sales Admin xác nhận Type of Lead       0
Sales                                   0
Số lần tương tác                        0
Trạng thái                              0
Level                                   0
Close Date                              0
Tỉnh/TP                                 0
Số lượng bộ sách                     2411
Số tiền giảm giá                     2414
Tổng tiền                               0
dtype: int64

In [ ]:
df2.head(5)

date                 Giờ  Khách hàng          SĐT  Channel  Chiến dịch  \
0  1-7-2020 2023-07-01 00:13:42        4713  123400000.0        0           1   
1  1-7-2020 2023-07-01 00:17:53        1143  123400001.0        0           1   
2  1-7-2020 2023-07-01 02:16:56        5143  123400002.0        0           1   
3  1-7-2020 2023-07-01 05:21:32        1829  123400003.0        0           1   
4  1-7-2020 2023-07-01 05:21:31        2537  123400004.0        0           1   

   Content  Marketer 2 Type of Lead Sales Admin xác nhận Type of Lead  \
0       51           1      Dathang                          DATTHANG   
1      119           9      Dathang                          DATTHANG   
2      240           9      Dathang                          DATTHANG   
3      119           9      Dathang                          DATTHANG   
4      135          16      Dathang                          DATTHANG   

      Sales  Số lần tương tác            Trạng thái Level  \
0  ThuyTran               1.0  Hủy đơn sau khi chốt  L7.2   
1  ThuyTran               1.0              Đặt hàng  L8.2   
2  ThuyTran               1.0  Hủy đơn sau khi chốt  L7.2   
3  ThuyTran               1.0  Hủy đơn sau khi chốt  L7.2   
4  ThuyTran               1.0              Đặt hàng  L8.2   

            Close Date     Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  \
0                    0      Hà Nội               1.0          140000.0   
1  2023-07-04 00:00:00  Quảng Ninh               1.0          140000.0   
2                    0      Hà Nội               1.0          140000.0   
3                    0      Hà Nội               1.0          140000.0   
4  2023-07-05 00:00:00      Hà Nội               1.0          140000.0   

   Tổng tiền  
0     499000  
1     499000  
2     499000  
3     499000  
4     499000

In [ ]:
for col in feature_cols:
    if col in encoders:  # Kiểm tra xem cột có trong danh sách mã hóa không
        df2[col] = encoders[col].inverse_transform(df2[col])


# Ta nhận thấy phần lớn các dòng só lượng bộ sách bih thiếu là do những trạng thái không phải là đặt hàng nên không thể có bộ sách nào nếu k đặt hàng

# Ta tiến hành sử dụng các điều kiện để đặt số lượng bộ sách nếu không phỉ tráng thái đặt hàng thì sẽ là 0

In [ ]:
df2.loc[(df2['Trạng thái'] != 'Đặt hàng') & (df2['Số lượng bộ sách'].isna()), 'Số lượng bộ sách'] = 0


Tương tự với số tiền giảm giá

In [ ]:
df2.loc[(df2['Trạng thái'] != 'Đặt hàng') & (df2['Số tiền giảm giá'].isna()), 'Số tiền giảm giá'] = 0

In [ ]:
df2.isnull().sum()

date                                 0
Giờ                                  0
Khách hàng                           0
SĐT                                  0
Channel                              0
Chiến dịch                           0
Content                              0
Marketer 2                           0
Type of Lead                         0
Sales Admin xác nhận Type of Lead    0
Sales                                0
Số lần tương tác                     0
Trạng thái                           0
Level                                0
Close Date                           0
Tỉnh/TP                              0
Số lượng bộ sách                     2
Số tiền giảm giá                     3
Tổng tiền                            0
dtype: int64

# Do 2 cột thiếu còn lại còn rất ít dòng bị thiếu( cụ thể là 2 và 3) nên ta tiến hành luôn xóa số dòng bị thiếu

In [ ]:
df2 = df2.dropna(subset=['Số lượng bộ sách','Số tiền giảm giá'])

# Xử lí trùng lặp( ơ đây khôbg có trùng lặp nên không cần xử lí)

In [ ]:
duplicate_rows = df2.duplicated()
df2[duplicate_rows]

Empty DataFrame
Columns: [date, Giờ, Khách hàng, SĐT, Channel, Chiến dịch, Content, Marketer 2, Type of Lead, Sales Admin xác nhận Type of Lead, Sales, Số lần tương tác, Trạng thái, Level, Close Date, Tỉnh/TP, Số lượng bộ sách, Số tiền giảm giá, Tổng tiền]
Index: []

In [ ]:
duplicates_count = df2.duplicated(subset=['Khách hàng', 'SĐT', 'Marketer 2']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")



Số dòng trùng lặp: 0


# III. Van Don

# Ta tiến hành show bảng 3

In [ ]:
df3.head(5)

STT Mã đơn hàng                                   Ghi chú đơn hàng  \
0  1.0    SON25144                                                NaN   
1  NaN         NaN                                                NaN   
2  2.0    SON24883  TẶNG SÁCH LẺ PHẦN 2 CHO PHỤ HUỲNH\ngiao trong ...   
3  NaN         NaN                                                NaN   
4  3.0    SON24943                          Giao trong giờ hành chính   

  Tags đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
0           NaN  Nguyễn Thị Huyền  Chi nhánh mặc định   Web       G87L96F3   
1           NaN               NaN                 NaN   NaN            NaN   
2           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web       G87L9VTM   
3           NaN               NaN                 NaN   NaN            NaN   
4           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web  NSAPOAAAAZPQF   

  Tình trạng gói hàng Trạng thái đối tác  ...  \
0        Đã giao hàng    Giao thành công  ...   
1                 NaN                NaN  ...   
2        Đã giao hàng    Giao thành công  ...   
3                 NaN                NaN  ...   
4        Đã giao hàng    Giao thành công  ...   

                                        Tên sản phẩm Ghi chú sản phẩm  \
0                                   Sách TeenUp - PH              NaN   
1                               BỘ SÁCH TEENUP - KID              NaN   
2                               Sách TeenUp - Phần 2              NaN   
3                              BỘ SÁCH TEENUP - TEEN              NaN   
4  COMBO KID - BỘ SÁCH TEENUP - 90 NGÀY DẬY THÌ T...              NaN   

   Số lượng Serial Đơn vị tính Đơn giá CK sản phẩm  CK tổng đơn hàng  \
0         1    NaN         NaN       0           0                 0   
1         1    NaN         NaN  499000           0                 0   
2         1    NaN         NaN       0           0                 0   
3         1    NaN          Bộ  639000      140000                 0   
4         1    NaN         NaN  920000      140000                 0   

  Thuế cho từng sản phẩm Tổng tiền hàng  
0                      0              0  
1                      0         499000  
2                      0              0  
3                      0         499000  
4                      0         780000  

[5 rows x 45 columns]

# Kiểm tra dữ liệu thiếu

In [ ]:
df3.isnull().sum()

STT                            2124
Mã đơn hàng                    2124
Ghi chú đơn hàng               7378
Tags đơn hàng                  7917
Nhân viên tạo đơn              2124
Chi nhánh                      2124
Nguồn                          2124
Mã vận đơn                     2124
Tình trạng gói hàng            2124
Trạng thái đối tác             2125
Lý do hủy đơn                  7944
Ngày đóng gói                  2124
Ngày hẹn giao                  7944
Ngày xuất kho                  2124
Ngày giao hàng                 2829
Đối tác giao hàng              2124
Dịch vụ giao hàng              2125
Khối lượng(g)                  2124
Kích thước(DxRxC)              2124
Tên người nhận                 2124
SĐT người nhận                 2124
Địa chỉ giao hàng              2124
Tỉnh/Thành                     2124
Quận/Huyện                     2124
Phường xã                      2124
Trạng thái đối soát            2124
Tiền khách phải trả cho đơn    2124
Khách hàng đã trả              2911
Hình thức thanh toán           2911
Tổng tiền thu hộ               2124
Phí vận chuyển                 2124
Người trả phí                  2124
Phí trả đối tác                2124
Ghi chú đơn giao               2124
Mã sản phẩm                       0
Tên sản phẩm                      0
Ghi chú sản phẩm               7944
Số lượng                          0
Serial                         7944
Đơn vị tính                    5646
Đơn giá                           0
CK sản phẩm                       0
CK tổng đơn hàng                  0
Thuế cho từng sản phẩm            0
Tổng tiền hàng                    0
dtype: int64

# Em tiến hành kiểm tra dữ liệu thiếu ở các côt, nếu cột nào có dữ liệu thiếu quá 50% thì em tiến hành xóa cột đđó sau xóa còn 17 cột

In [ ]:
# Xác định ngưỡng 50% số hàng
threshold = len(df3) * 0.5

# Xóa cột có hơn 50% giá trị bị thiếu
df3 = df3.dropna(axis=1, thresh=threshold)

# em tiến hành xóa những hàng thừa ở những cột STT, Trạng thái đối tác, dịch vụ giao hàng vì đó là những dòng trắng không cần thiết( em có xem ở exel)

In [ ]:
df3= df3.dropna(subset=['STT','Trạng thái đối tác','Dịch vụ giao hàng'])

In [ ]:
df3.isnull().sum()

STT                              0
Mã đơn hàng                      0
Nhân viên tạo đơn                0
Chi nhánh                        0
Nguồn                            0
Mã vận đơn                       0
Tình trạng gói hàng              0
Trạng thái đối tác               0
Ngày đóng gói                    0
Ngày xuất kho                    0
Ngày giao hàng                 705
Đối tác giao hàng                0
Dịch vụ giao hàng                0
Khối lượng(g)                    0
Kích thước(DxRxC)                0
Tên người nhận                   0
SĐT người nhận                   0
Địa chỉ giao hàng                0
Tỉnh/Thành                       0
Quận/Huyện                       0
Phường xã                        0
Trạng thái đối soát              0
Tiền khách phải trả cho đơn      0
Khách hàng đã trả              787
Hình thức thanh toán           787
Tổng tiền thu hộ                 0
Phí vận chuyển                   0
Người trả phí                    0
Phí trả đối tác                  0
Ghi chú đơn giao                 0
Mã sản phẩm                      0
Tên sản phẩm                     0
Số lượng                         0
Đơn giá                          0
CK sản phẩm                      0
CK tổng đơn hàng                 0
Thuế cho từng sản phẩm           0
Tổng tiền hàng                   0
dtype: int64

In [ ]:
df3= df3.drop(columns=["Ngày giao hàng"])


# Em áp dụng phương pháp học máy để dự đoán và điền giá trị bị thiêys ở cột khách hàng đã trả

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Khách hàng đã trả']  # Các cột cần dự đoán
feature_cols = ["STT", "Mã đơn hàng", "Chi nhánh", "Mã sản phẩm","Khối lượng(g)"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df3[col] = le.fit_transform(df3[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần, sử dụng df3 thay vì df2
    df_cleaned_train = df3[df3[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df3[df3[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df3 với giá trị đã điền, sử dụng df3 thay vì df2
    df3.loc[df3[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df3[columns_missing].isnull().sum())  # Kiểm tra còn NaN không

Đang xử lý cột: Khách hàng đã trả
Số giá trị NaN sau khi xử lý:
Khách hàng đã trả    0
dtype: int64


<ipython-input-79-4eaa760f3cb5>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


# Tương tự với cột Hình thức thanh toán

In [ ]:
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder

columns_missing = ['Hình thức thanh toán']  # Các cột cần dự đoán
feature_cols = ["STT", "Mã đơn hàng", "Chi nhánh", "Mã sản phẩm","Khách hàng đã trả"]  # Các cột đặc trưng

# Mã hóa tất cả các cột đặc trưng thành số
encoders = {}
for col in feature_cols:
    le = LabelEncoder()
    df3[col] = le.fit_transform(df3[col].astype(str))  # Chuyển đổi cột thành số
    encoders[col] = le  # Lưu encoder để sử dụng sau

# Xử lý từng cột bị thiếu dữ liệu
for col_missing in columns_missing:
    print(f"Đang xử lý cột: {col_missing}")

    # Chia dữ liệu thành hai phần, sử dụng df3 thay vì df2
    df_cleaned_train = df3[df3[col_missing].notnull()]  # Dữ liệu đầy đủ
    df_cleaned_missing = df3[df3[col_missing].isnull()]  # Dữ liệu thiếu

    # Dữ liệu đầu vào (đã được mã hóa)
    X_train = df_cleaned_train[feature_cols]
    y_train = df_cleaned_train[col_missing]

    # Mã hóa cột cần dự đoán
    label_encoder = LabelEncoder()
    y_train_encoded = label_encoder.fit_transform(y_train.astype(str))

    # Huấn luyện mô hình Decision Tree
    model = DecisionTreeClassifier(max_depth=10, random_state=42)
    model.fit(X_train, y_train_encoded)

    # Dự đoán giá trị bị thiếu
    X_missing = df_cleaned_missing[feature_cols]
    predicted_encoded_values = model.predict(X_missing)

    # Giải mã từ số về dạng chữ
    df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)

    # Cập nhật lại df3 với giá trị đã điền, sử dụng df3 thay vì df2
    df3.loc[df3[col_missing].isnull(), col_missing] = df_cleaned_missing[col_missing]

# Kiểm tra kết quả
print("Số giá trị NaN sau khi xử lý:")
print(df3[columns_missing].isnull().sum())  # Kiểm tra còn Na

Đang xử lý cột: Hình thức thanh toán
Số giá trị NaN sau khi xử lý:
Hình thức thanh toán    0
dtype: int64


<ipython-input-80-c743db58f585>:41: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned_missing[col_missing] = label_encoder.inverse_transform(predicted_encoded_values)


In [ ]:
df3.isnull().sum()

STT                            0
Mã đơn hàng                    0
Nhân viên tạo đơn              0
Chi nhánh                      0
Nguồn                          0
Mã vận đơn                     0
Tình trạng gói hàng            0
Trạng thái đối tác             0
Ngày đóng gói                  0
Ngày xuất kho                  0
Đối tác giao hàng              0
Dịch vụ giao hàng              0
Khối lượng(g)                  0
Kích thước(DxRxC)              0
Tên người nhận                 0
SĐT người nhận                 0
Địa chỉ giao hàng              0
Tỉnh/Thành                     0
Quận/Huyện                     0
Phường xã                      0
Trạng thái đối soát            0
Tiền khách phải trả cho đơn    0
Khách hàng đã trả              0
Hình thức thanh toán           0
Tổng tiền thu hộ               0
Phí vận chuyển                 0
Người trả phí                  0
Phí trả đối tác                0
Ghi chú đơn giao               0
Mã sản phẩm                    0
Tên sản phẩm                   0
Số lượng                       0
Đơn giá                        0
CK sản phẩm                    0
CK tổng đơn hàng               0
Thuế cho từng sản phẩm         0
Tổng tiền hàng                 0
dtype: int64

# Kiem tra trung lap

In [ ]:
duplicate_rows = df3.duplicated()
df3[duplicate_rows]

Empty DataFrame
Columns: [STT, Mã đơn hàng, Nhân viên tạo đơn, Chi nhánh, Nguồn, Mã vận đơn, Tình trạng gói hàng, Trạng thái đối tác, Ngày đóng gói, Ngày xuất kho, Đối tác giao hàng, Dịch vụ giao hàng, Khối lượng(g), Kích thước(DxRxC), Tên người nhận, SĐT người nhận, Địa chỉ giao hàng, Tỉnh/Thành, Quận/Huyện, Phường xã, Trạng thái đối soát, Tiền khách phải trả cho đơn, Khách hàng đã trả, Hình thức thanh toán, Tổng tiền thu hộ, Phí vận chuyển, Người trả phí, Phí trả đối tác, Ghi chú đơn giao, Mã sản phẩm, Tên sản phẩm, Số lượng, Đơn giá, CK sản phẩm, CK tổng đơn hàng, Thuế cho từng sản phẩm, Tổng tiền hàng]
Index: []

[0 rows x 37 columns]

In [ ]:
duplicates_count = df3.duplicated(subset=['Mã đơn hàng', 'Mã sản phẩm']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 1


In [ ]:
for col in feature_cols:
    if col in encoders:  # Kiểm tra xem cột có trong danh sách mã hóa không
        df3[col] = encoders[col].inverse_transform(df3[col])


In [ ]:
df3 = df3.drop_duplicates(subset=['Mã đơn hàng', 'Mã sản phẩm'], keep='first')

In [ ]:
duplicates_count = df3.duplicated(subset=['Mã đơn hàng', 'Mã sản phẩm']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 0


In [ ]:
duplicates_count = df3.duplicated(subset=['Mã đơn hàng', 'Mã sản phẩm','Mã vận đơn']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 0


In [ ]:
duplicates_count = df3.duplicated(subset=['Mã vận đơn']).sum()
print(f"Số dòng trùng lặp: {duplicates_count}")


Số dòng trùng lặp: 0


In [ ]:
!pip install xlsxwriter # Install the xlsxwriter package if not installed
import pandas as pd

# Đường dẫn lưu file đã xử lý (cùng thư mục với file gốc)
output_drive_path = '/content/drive/My Drive/Technical_Test/Database_Q3_2020_Processed.xlsx'

# Lưu file với nhiều sheet
with pd.ExcelWriter(output_drive_path, engine='xlsxwriter') as writer: # Using xlsxwriter to handle excel files
    df1.to_excel(writer, sheet_name="Sheet1", index=False)
    df2.to_excel(writer, sheet_name="Sheet2", index=False)
    df3.to_excel(writer, sheet_name="Sheet3", index=False)

print(f"File đã được lưu vào Google Drive: {output_drive_path}")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 165.1/165.1 kB 3.5 MB/s eta 0:00:00
File đã được lưu vào Google Drive: /content/drive/My Drive/Technical_Test/Database_Q3_2020_Processed.xlsx
